### PR TITLE
Adopt the ArgumentParser library for the command-line tools

### PR DESCRIPTION
**Pull request by natecook1000**
This transitions the different SPM command-line tools to use the `ArgumentParser` library instead of the TSC parser. Among the changes is that the tools are no longer subclasses of `SwiftTool`. Instead, each tool creates a `SwiftTool` instance to use when necessary. It may be useful to follow up with a refactoring to either move some of the functionality that's in the tools into `SwiftTool` or to otherwise separate the CLI from the actual functions being performed.

**Still to do:**
- [x] Get tests passing
- [x] Add shell completion generation to `ArgumentParser`

### CODE REVIEW

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@swift-ci please smoke test self hosted

**Comment by natecook1000** [comment]
@swift-ci Please smoke test self hosted

**Comment by natecook1000** [comment]
@swift-ci Please smoke test self hosted

**Comment by natecook1000** [comment]
@swift-ci Please smoke test self hosted

**Comment by natecook1000** [comment]
@swift-ci Please smoke test self hosted

**Comment by natecook1000** [comment]
@neonichu @abertelrud With shell completions now included in `ArgumentParser`, this is now ready for review.

**Comment by natecook1000** [comment]
@swift-ci Please smoke test self hosted

File: Tests/CommandsTests/PackageToolTests.swift [hunk 1]
```diff
@@ -727,7 +727,7 @@ final class PackageToolTests: XCTestCase {
                 try execute(["config", "get-mirror", "--original-url", "git@github.com:apple/swift-package-manager.git"], packagePath: packageRoot)
             }
 
-            check(stderr: "error: mirror not found\n") {
+            check(stderr: "Error: mirror not found\n") {
```
Comment (aciidb0mb3r) [inline Tests/CommandsTests/PackageToolTests.swift:730]: Command-line swiftpm tries to follow compiler like diagnostics so this diff is unexpected.

Comment (natecook1000) [reply Tests/CommandsTests/PackageToolTests.swift]: Updated to match the existing diagnostics, thanks!

**Comment by natecook1000** [comment]
@swift-ci Please smoke test self hosted

**Comment by tomerd** [comment]
hey @natecook1000 we would love to get this merged, @abertelrud will work with you to get this done

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

File: Package.swift [hunk 2]
```diff
@@ -72,6 +72,11 @@ let package = Package(
             targets: ["PackageDescription"]
         ),
     ],
+    dependencies: [
```
Comment (abertelrud) [inline Package.swift:75]: We're probably going to need to handle these dependencies as we do with the others, at the bottom of the manifest.  The reason is that in some cases the build needs to be run with all the dependencies checked out next to SwiftPM.  I will take a look at what's needed here based on how it's being built.

**Comment by abertelrud** [comment]
At this point this is just waiting for the CI to be configured to check out 0.3.0 of ArgumentParser (those are the PRs that @owenv mentioned; they just need to be approved and merged, then we can merge this).

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by abertelrud** [comment]
The failure is unrelated (`error: 'Triple' is ambiguous for type lookup in this context`) and is being worked on, but we'll need to wait to merge until it's fixed.

**Comment by abertelrud** [comment]
Now that https://github.com/apple/swift-package-manager/pull/2883 is merged, we can try this CI again.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
The Linux failure seems to involve an error in SourceKit-LSP's `build-script-helper.py` script.  Need to investigate whether it's related.

**Comment by abertelrud** [comment]
It does look related:

15:13:49 + /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/sourcekit-lsp/Utilities/build-script-helper.py build --package-path /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/sourcekit-lsp --build-path /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/build/buildbot_incremental/sourcekitlsp-linux-x86_64 --configuration release --toolchain /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/build/buildbot_incremental/toolchain-linux-x86_64/usr --ninja-bin /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/build/buildbot_incremental/ninja-build/ninja --verbose
15:13:49 error: Missing value for '-c <configuration>'

So this might be a difference in how this flag is being parsed.

**Comment by abertelrud** [comment]
I don't see any problems with how the `build-script-helper.py` script is forming the arguments.  Something odd is going on, though: the script always passes `--configuration`, but the error message is about `-c`.  In the error message, I always see `swift-build` echoing back the form (log or short) of the flag that was actually passed to it, it's possible that it's a different invocation than then one directly in `build-script-helper.py` and that the lines are just intermixed.

**Comment by abertelrud** [comment]
I can actually reproduce this using the exact line that SourceKit-LSP is passing:
```
.build/debug/swift-build  --package-path /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/sourcekit-lsp --build-path /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/build/buildbot_incremental/sourcekitlsp-linux-x86_64 --configuration release --verbose -Xcxx -I -Xcxx /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/build/buildbot_incremental/toolchain-linux-x86_64/usr/lib/swift -Xcxx -I -Xcxx /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test/branch-master/build/buildbot_incremental/toolchain-linux-x86_64/usr/lib/swift/Block -Xlinker -rpath -Xlinker ../lib/swift/linux` 
error: Missing value for '-c <configuration>'
```
with the changes in this PR, but it doesn't reproduce without them.  I'll see if I can figure out what's causing it to think it's seeing a `-c`.

**Comment by natecook1000** [comment]
Looks like the C++ compiler flag option had an incorrect name. Since `-Xcxx` wasn't matching as a full name, the parser was treating it as a group of short options, and recognized `-c`, but expected a value to follow. Better diagnostics would help here!

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by abertelrud** [comment]
Fantastic; thank you!  I guess this wasn't seen on macOS because the script only passes that flag on Linux?  Not sure what happened there.  I guess I also wonder whether there are other combinations of flags that might get passed and that might be misinterpreted as combined single-file flags, resulting in confusing error messages.  Improving handling of incorrectly specified parameters should not prevent us from merging, of course, but I am wondering whether we should disable combining of short-from flags for SwiftPM.

**Comment by abertelrud** [comment]
@natecook1000, assuming this passes, are you fine with merging it?  I think all the other issues have been taken care of (downstream CI etc).

**Pull request merged**